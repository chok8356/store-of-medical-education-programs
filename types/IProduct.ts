import { ICategory } from './ICategory'

export interface IProduct {
  id: number;
  name: string;
  img: string;
  price: number;
  description: string
  category: ICategory['id']
}

export interface IProductFilters {
  query: IProduct['name'] | IProduct['description'],
  selectedCategory: ICategory['id'] | null
}
