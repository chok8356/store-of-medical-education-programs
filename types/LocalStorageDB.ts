export class LocalStorageDB {
  dbPrefix = 'store@'

  constructor() {
    const instance = (this.constructor as any).instance
    if (instance) {
      return instance
    }
    (this.constructor as any).instance = this
  }

  private getPrefix(key: string) {
    return this.dbPrefix + key
  }

  get(key: string) {
    const data = localStorage.getItem(this.getPrefix(key))
    return data ? JSON.parse(data) : undefined
  }

  put(key: string, value: any) {
    const _value = JSON.stringify(value)
    localStorage.setItem(this.getPrefix(key), _value)
    return true
  }

  delete(key: string) {
    localStorage.removeItem(this.getPrefix(key))
  }
}
