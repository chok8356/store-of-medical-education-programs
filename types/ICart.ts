import { IProduct } from './IProduct'

export interface ICart {
  [id: number]: ICartItem
}

export interface ICartItem extends Pick<IProduct, 'id' | 'name' | 'price'> {
  quantity: number;
}
