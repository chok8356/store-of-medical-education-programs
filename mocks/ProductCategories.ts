import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
