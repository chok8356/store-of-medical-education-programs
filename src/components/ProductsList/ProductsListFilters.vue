<template>
  <div
    v-if="filters"
    class="products-list-filters card p-4 shadow-sm">
    <b-form-group
      class="products-list-filters__search"
      label="Программы НМО">
      <b-form-input
        placeholder="Поиск"
        type="text"
        :value="filters.query"
        @input="onQueryInput" />
    </b-form-group>

    <hr class="mt-3 mb-4">

    <div class="products-list-filters__categories">
      <h5 class="d-block mb-3">
        Категории
      </h5>
      <b-nav
        pills
        vertical>
        <b-nav-item
          :active="filters.selectedCategory === null"
          @click="onCategoryClick(null)">
          Все категории
        </b-nav-item>
        <b-nav-item
          v-for="category in categories"
          :key="category.id"
          :active="filters.selectedCategory === category.id"
          @click="onCategoryClick(category.id)">
          {{ category.name }}
        </b-nav-item>
      </b-nav>
    </div>
  </div>
</template>

<script lang="ts">
import { BNav, BNavItem, BFormGroup, BFormInput } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop, Model } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProductFilters } from '../../../types/IProduct'

@Component({
  name: 'ProductsListFilters',
  components: {
    BNav, BNavItem, BFormGroup, BFormInput
  }
})
export default class ProductsListFilters extends Vue {
  @Model('change', { type: Object as PropType<IProductFilters> }) filters!: IProductFilters
  @Prop({ type: Array as PropType<ICategory[]> }) categories!: ICategory[]

  onCategoryClick(id: ICategory['id'] | null) {
    this.$emit('change', {
      ...this.filters,
      selectedCategory: id
    })
  }

  onQueryInput(value: string) {
    this.$emit('change', {
      ...this.filters,
      query: value
    })
  }
}
</script>

<style lang="scss">
.products-list-filters {
  position: sticky;
  top: 104px;
  z-index: 1;
}
</style>
