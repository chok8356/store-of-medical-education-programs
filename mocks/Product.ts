import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
