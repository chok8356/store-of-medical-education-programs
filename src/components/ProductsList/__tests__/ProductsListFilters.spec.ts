import { shallowMount } from '@vue/test-utils'

import { BFormInput, BNavItem } from 'bootstrap-vue'
import { categories } from '@/../mocks/ProductCategories'
import { IProductFilters } from '../../../../types/IProduct'
import ProductsListFilters from '../ProductsListFilters.vue'

const filters: IProductFilters = {
  query: '',
  selectedCategory: null
}

describe('ProductsListFilters', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(ProductsListFilters, {
      propsData: {
        filters,
        categories
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('exist props', () => {
    expect(wrapper.props('filters')).toStrictEqual(filters)
    expect(wrapper.props('categories')).toStrictEqual(categories)
  })

  it('change query on input', () => {
    const inputComponent = wrapper.findComponent(BFormInput)
    inputComponent.vm.$emit('input', '1')
    expect(wrapper.emitted('change')[0][0]).toStrictEqual({
      query: '1',
      selectedCategory: null
    })
  })

  it('change selected category on category click', () => {
    const categoryComponent = wrapper.findAllComponents(BNavItem).at(1)
    categoryComponent.trigger('click')
    expect(wrapper.emitted('change')[0][0]).toStrictEqual({
      query: '',
      selectedCategory: categories[0].id
    })
  })
})
