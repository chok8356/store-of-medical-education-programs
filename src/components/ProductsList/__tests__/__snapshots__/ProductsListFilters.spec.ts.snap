// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsListFilters render correctly 1`] = `
<div
  class="products-list-filters card p-4 shadow-sm"
>
  <b-form-group-stub
    class="products-list-filters__search"
    feedbackarialive="assertive"
    label="Программы НМО"
  >
    <b-form-input-stub
      debounce="0"
      formatter="[Function]"
      placeholder="Поиск"
      type="text"
      value=""
    />
  </b-form-group-stub>
   
  <hr
    class="mt-3 mb-4"
  />
   
  <div
    class="products-list-filters__categories"
  >
    <h5
      class="d-block mb-3"
    >
      
      Категории
    
    </h5>
     
    <b-nav-stub
      pills="true"
      tag="ul"
      vertical="true"
    >
      <b-nav-item-stub
        active="true"
        linkattrs="[object Object]"
        target="_self"
      >
        
        Все категории
      
      </b-nav-item-stub>
       
      <b-nav-item-stub
        linkattrs="[object Object]"
        target="_self"
      >
        
        Категория 1
      
      </b-nav-item-stub>
      <b-nav-item-stub
        linkattrs="[object Object]"
        target="_self"
      >
        
        Категория 2
      
      </b-nav-item-stub>
      <b-nav-item-stub
        linkattrs="[object Object]"
        target="_self"
      >
        
        Категория 3
      
      </b-nav-item-stub>
      <b-nav-item-stub
        linkattrs="[object Object]"
        target="_self"
      >
        
        Категория 4
      
      </b-nav-item-stub>
      <b-nav-item-stub
        linkattrs="[object Object]"
        target="_self"
      >
        
        Категория 5
      
      </b-nav-item-stub>
    </b-nav-stub>
  </div>
</div>
`;
